### DISCUSSION TITLE
What is the best practice to use cypress for testing?

### DISCUSSION THREAD

**Discussion by dev1000-youtuber** [cypress-io/cypress#16668] [CI setup]
Question 1: Should we run tests on all branches? Or just on master branch?

Question 2: When should the cypress tests run? On each commit or there is other approach?

**Comment by sandra-ouadghiri** [comment]
Hi,
It seems more like a pipeline & test strategy than a Cypress matter. The same could be discussed on another test framework.
This said, here's my opinion. Hopefully this will help you.

TLDR;
- Sooner is better (shift left)
- But testing early implies stuff your team may not be ready for

Sooner is better (shift left).
Testing your main brnach before you release to your client is the basic. But if red, you still need to 'hotfix' your main branch, and people may need to rebase not to propagate. This leads to regressions being reintroduced by mistakes or conflicts issues: so not great. 'Obviously' it is nicer & less costly for everyone to spot a bug before merge than after, especially if automation can catch it -> so you should test before merge if doable.

But
This said, this means that for n branches (or n PR/MR) you need n environments (or let's say 'independant reachable product', it may be 'real' environment or maybe something less extravagent) to run your tests on.
Your pipeline has triggers. If the trigger is 'every time someone pushes' then this should trigger test executions on env xyz. Maybe this implies deploying an env first. How long does it take for your team? If 1h then you have a problem. If you don't know how to deploy a test env, then you have a problem. And thus are stucked to 'test develop HEAD on test1 env every 2h if something was merged'.
Now let's say deployment is not a problem, then how long does the test execution takes? If 5mn, fine, but if 4h, you can't expect a dev to wait 4h for a push result. Especially if they push regularly.

Now dev can also 'just' keep cypress around as they code, and run all tests from time to time on their local env (and before asking for PR review). This still implies that: your devs all have a local env. That they know how to launch Cypress locally (and have access to Cypress code). And that once again, test execution remains small.

If you wonder what is 'classical' on a company: in my experience I get hired because test runs only on develop at night (nor not even) and companies (and devs) wants it more often & earlier ie on their branch at push or PR opening/merging. It's our job to find ways to make that possible, with tremendous help from TechOps people ❤️.
